File: packages/console-core/src/tools.test.ts
import { readdir, readFile } from "node:fs/promises";
import { describe, expect, it } from "vitest";
import {
  INTERNAL_TOOLS,
  TOOL_GROUPS,
  toolUrl,
  toolsInGroup,
  type ToolGroup,
} from "./tools";

describe("the tool list is data, not markup", () => {
  it("derives URLs from the environment's base domain", () => {
    const zitadel = INTERNAL_TOOLS.find((t) => t.name === "Zitadel");
    expect(zitadel).toBeDefined();
    expect(toolUrl(zitadel!, "tesserix.app")).toBe("https://auth.tesserix.app");
    // The point of deriving: a non-production console must not link operators
    // at production tools.
    expect(toolUrl(zitadel!, "dev.tesserix.app")).toBe(
      "https://auth.dev.tesserix.app",
    );
  });

  it("hardcodes no absolute URLs", () => {
    // A full URL in the data would defeat the environment awareness above and
    // would not fail visibly — it would just always point at prod.
    for (const tool of INTERNAL_TOOLS) {
      expect(tool.subdomain, `${tool.name} looks like a full URL`).not.toMatch(
        /^https?:|\./,
      );
    }
  });

  it("has no duplicate subdomains", () => {
    // Still asserted, but it is no longer the only guarantee: platform_tools has
    // a UNIQUE constraint on subdomain (migration 0031). This keeps the FALLBACK
    // list honest — it is what renders when the platform API cannot be reached.
    const seen = INTERNAL_TOOLS.map((t) => t.subdomain);
    expect(new Set(seen).size).toBe(seen.length);
  });

  it("holds exactly the directory the migrations describe", async () => {
    // The literal and the migrations are two copies of one directory. This is
    // the only thing that keeps them equal (#499).
    //
    // It replaces a one-directional check that read 0031 alone and asked
    // whether each fallback subdomain appeared ANYWHERE in the file. That
    // missed both ways drift actually happens:
    //
    //   - a tool added by migration and not here — the fallback then serves a
    //     directory shorter than the live one during an outage, silently;
    //   - a tool DELETED by migration and left here — 0042 removed
    //     `secret-service`, and a `toContain` against 0031 still passes for it
    //     forever, because 0031 does still mention it. That is exactly the
    //     drift #486 introduced in the other copy and nobody noticed.
    //
    // So the expectation is computed the way the database computes it: every
    // seeded row, minus every deleted one, in migration order.
    const expected = await directoryFromMigrations();

    // Compared as sorted `subdomain → name` lines rather than as Maps: the
    // failure then names the tool that drifted instead of printing two maps.
    const fromMigrations = Array.from(
      expected,
      ([subdomain, name]) => `${subdomain} → ${name}`,
    ).sort();
    const fromFallback = INTERNAL_TOOLS.map(
      (t) => `${t.subdomain} → ${t.name}`,
    ).sort();

    expect(
      fromFallback,
      "the fallback and the migrations describe different directories",
    ).toEqual(fromMigrations);
  });
});

/**
 * The tools directory as the migrations leave it: `subdomain -> name`.
 *
 * Reads every migration rather than a named pair, so a future 0050 that adds
 * or removes a tool is picked up without editing this test — the failure mode
 * this guards against is precisely someone changing the directory and not
 * thinking about the second copy.
 */
async function directoryFromMigrations(): Promise<Map<string, string>> {
  const dir = new URL("../../../apps/web/db/migrations/", import.meta.url);
  const files = (await readdir(dir)).filter((f) => f.endsWith(".sql")).sort();

  const directory = new Map<string, string>();
  for (const file of files) {
    const sql = await readFile(new URL(file, dir), "utf8");
    if (!sql.includes("platform_tools")) continue;

    for (const statement of insertsInto(sql, "platform_tools")) {
      // ('Name', 'subdomain', 'purpose', … — name and subdomain are the two
      // leading literals of each VALUES row and neither may contain a quote.
      const rows = Array.from(
        statement.matchAll(/\(\s*'([^']+)',\s*'([^']+)'/g),
      );
      for (const [, name, subdomain] of rows) {
        directory.set(subdomain, name);
      }
    }

    const deletions = Array.from(
      sql.matchAll(
        /DELETE\s+FROM\s+platform_tools\s+WHERE\s+subdomain\s*=\s*'([^']+)'/gi,
      ),
    );
    for (const [, subdomain] of deletions) {
      directory.delete(subdomain);
    }
  }

  // A directory that parsed to nothing would make every assertion above vacuous
  // — the classic way a regex-driven test goes quietly green after a schema
  // rename.
  expect(directory.size, "parsed no tools out of the migrations").toBeGreaterThan(0);
  return directory;
}

/**
 * The body of each `INSERT INTO <table> … VALUES …;` statement in a file.
 *
 * The statement ends at a semicolon that ENDS ITS LINE, not at the first
 * semicolon. 0031 seeds the note "Reached outside the Istio gateway; its own
 * login." — splitting on the first `;` truncates the statement mid-row and
 * silently drops the six tools after it, which is a green test asserting
 * almost nothing.
 */
function insertsInto(sql: string, table: string): string[] {
  const bodies: string[] = [];
  const opening = new RegExp(`INSERT\\s+INTO\\s+${table}\\s*\\(`, "gi");
  for (const match of Array.from(sql.matchAll(opening))) {
    const rest = sql.slice(match.index);
    const end = /;[ \t]*(?:\r?\n|$)/.exec(rest);
    bodies.push(end ? rest.slice(0, end.index) : rest);
  }
  return bodies;
}

describe("grouping", () => {
  it("assigns every tool to a declared group", () => {
    // The live path's guarantee is now a foreign key with ON DELETE RESTRICT.
    // This asserts the same thing about the fallback, which no foreign key
    // covers.
    const declared = new Set<ToolGroup>(TOOL_GROUPS.map((g) => g.key));
    for (const tool of INTERNAL_TOOLS) {
      expect(declared, `${tool.name} has an undeclared group`).toContain(
        tool.group,
      );
    }
  });

  it("leaves no group empty", () => {
    // An empty group renders as a heading with nothing under it, which reads
    // as a loading failure rather than an absence.
    //
    // NOTE: with groups in platform_tool_groups this can no longer be
    // guaranteed for the LIVE directory — a foreign key cannot express "and at
    // least one tool references you". The renderer skips an empty group
    // (components/internal-tools.tsx) and internal-tools.render.test.tsx covers
    // that. What this test still guarantees is the fallback.
    for (const group of TOOL_GROUPS) {
      expect(
        toolsInGroup(group.key).length,
        `group "${group.label}" is declared but empty`,
      ).toBeGreaterThan(0);
    }
  });

  it("covers every tool exactly once across the groups", () => {
    const total = TOOL_GROUPS.reduce(
      (sum, g) => sum + toolsInGroup(g.key).length,
      0,
    );
    expect(total).toBe(INTERNAL_TOOLS.length);
  });
});

describe("entries say something useful", () => {
  it("gives every tool a purpose a newcomer could act on", () => {
    for (const tool of INTERNAL_TOOLS) {
      expect(tool.purpose.length, `${tool.name} has no purpose`).toBeGreaterThan(
        15,
      );
      // A directory whose descriptions restate the name teaches nothing.
      expect(
        tool.purpose.toLowerCase(),
        `${tool.name}'s purpose just restates its name`,
      ).not.toBe(tool.name.toLowerCase());
    }
  });

  it("records access notes only where there is something real to say", () => {
    // Guards against the field drifting into "unknown" everywhere, which would
    // be noise, or into guesses, which would be worse than silence.
    const noted = INTERNAL_TOOLS.filter((t) => t.note);
    expect(noted.length).toBeGreaterThan(0);
    expect(noted.length).toBeLessThan(INTERNAL_TOOLS.length / 2);
    for (const tool of noted) {
      expect(tool.note!.toLowerCase()).not.toContain("unknown");
    }
  });
});

File: platform-api/internal/modules/tools/internal/repository/tools_test.go
package repository_test

import (
	"context"
	"testing"

	"github.com/tesserix/tesserix-home/platform-api/internal/modules/tools/internal/domain"
	"github.com/tesserix/tesserix-home/platform-api/internal/modules/tools/internal/repository"
	"github.com/tesserix/tesserix-home/platform-api/internal/platform/testdb"
)

// The seed is part of the migration rather than a fixture, because the
// directory must render identically on the day of cutover. These assert the
// seed itself, so a migration that created the tables and forgot the rows
// fails here rather than on the home page.

func TestTheSeedCarriesTodaysDirectory(t *testing.T) {
	pool := testdb.New(t)
	ctx := context.Background()

	groups, err := repository.ListGroups(ctx, pool)
	if err != nil {
		t.Fatalf("listing groups: %v", err)
	}
	wantGroups := []string{"identity", "observability", "delivery", "cost", "reference"}
	if len(groups) != len(wantGroups) {
		t.Fatalf("got %d groups, want %d: %+v", len(groups), len(wantGroups), groups)
	}
	// Order is the contract: TOOL_GROUPS is a display order, and identity is
	// first because it gates everything else.
	for i, want := range wantGroups {
		if groups[i].Key != want {
			t.Errorf("group %d = %q, want %q — sort_order is the display order", i, groups[i].Key, want)
		}
	}

	tools, err := repository.ListTools(ctx, pool)
	if err != nil {
		t.Fatalf("listing tools: %v", err)
	}
	if len(tools) != 14 {
		t.Fatalf("got %d tools, want the 14 in tools.ts: %+v", len(tools), tools)
	}
	if tools[0].Subdomain != "auth" || tools[0].Name != "Zitadel" {
		t.Errorf("first tool = %s/%s, want Zitadel/auth", tools[0].Name, tools[0].Subdomain)
	}
}

func TestTheOneToolWithANoteKeepsIt(t *testing.T) {
	pool := testdb.New(t)

	tools, err := repository.ListTools(context.Background(), pool)
	if err != nil {
		t.Fatalf("listing tools: %v", err)
	}

	noted := map[string]string{}
	for _, tool := range tools {
		if tool.Note != nil {
			noted[tool.Subdomain] = *tool.Note
		}
	}
	// Exactly the one that carries one in tools.ts. A note is present only
	// where there is something real to say, and the seed must not invent any.
	//
	// This was two until 0042 deleted the secret-service row, which held the
	// other one: the standalone secrets UI it described is gone, so its
	// "separate login" note describes nothing. argocd is the survivor.
	if len(noted) != 1 {
		t.Errorf("got %d notes, want 1 (argocd): %+v", len(noted), noted)
	}
	if _, ok := noted["argocd"]; !ok {
		t.Error("argocd lost its outside-the-gateway note")
	}
}

func TestASubdomainThatIsNotADnsLabelIsRefused(t *testing.T) {
	pool := testdb.New(t)
	ctx := context.Background()

	// The CHECK is what keeps toolUrl's host derivation honest. Without it a
	// row could carry a whole URL and link operators anywhere.
	for _, bad := range []string{
		"https://grafana.tesserix.app",
		"grafana.tesserix.app",
		"Grafana",
		"-leading",
		"trailing-",
		"",
	} {
		_, err := pool.Exec(ctx,
			`INSERT INTO platform_tools (name, subdomain, purpose, group_key, sort_order)
			 VALUES ('x', $1, 'x', 'reference', 99)`, bad)
		if err == nil {
			t.Errorf("subdomain %q was accepted; it must be a single DNS label", bad)
		}
	}
}

func TestAToolInAnUndeclaredGroupIsRefused(t *testing.T) {
	pool := testdb.New(t)

	// This is the foreign key standing in for the ToolGroup union type that
	// moving groups into the database gives up.
	_, err := pool.Exec(context.Background(),
		`INSERT INTO platform_tools (name, subdomain, purpose, group_key, sort_order)
		 VALUES ('x', 'x', 'x', 'no-such-group', 99)`)
	if err == nil {
		t.Fatal("a tool in an undeclared group was accepted; the FK is missing")
	}
}

func TestADuplicateSubdomainIsRefused(t *testing.T) {
	pool := testdb.New(t)

	_, err := pool.Exec(context.Background(),
		`INSERT INTO platform_tools (name, subdomain, purpose, group_key, sort_order)
		 VALUES ('Another Zitadel', 'auth', 'x', 'identity', 99)`)
	if err == nil {
		t.Fatal("a duplicate subdomain was accepted; the UNIQUE constraint is missing")
	}
}

func TestAGroupWithToolsCannotBeDeleted(t *testing.T) {
	pool := testdb.New(t)

	_, err := pool.Exec(context.Background(),
		`DELETE FROM platform_tool_groups WHERE key = 'identity'`)
	if err == nil {
		t.Fatal("a group with tools was deleted; ON DELETE RESTRICT is missing")
	}
}

func TestTheApiAndTheDatabaseRefuseTheSameSubdomains(t *testing.T) {
	pool := testdb.New(t)
	ctx := context.Background()

	// The regex is written twice — once in domain, once in migration 0031 —
	// and this is what keeps the copies honest. A divergence means either a
	// 500 from a constraint the API thought it had already checked, or a row
	// the API refuses that the database would have taken.
	for _, candidate := range []string{
		"a", "ab", "z9", "secret-service-2",
		"https://grafana.tesserix.app", "grafana.tesserix.app",
		"Grafana", "-leading", "trailing-", "under_score", "",
	} {
		apiOK := domain.Tool{Name: "x", Subdomain: candidate, Purpose: "x",
			GroupKey: "reference"}.Normalise().Validate() == nil

		_, err := pool.Exec(ctx,
			`INSERT INTO platform_tools (name, subdomain, purpose, group_key, sort_order)
			 VALUES ('probe', $1, 'probe', 'reference', 999)`, candidate)
		dbOK := err == nil
		if dbOK {
			_, _ = pool.Exec(ctx, `DELETE FROM platform_tools WHERE subdomain = $1 AND name = 'probe'`, candidate)
		}

		// "Grafana" is the one legitimate asymmetry: the API lower-cases it in
		// Normalise, so it accepts what the raw database refuses.
		if candidate == "Grafana" {
			continue
		}
		if apiOK != dbOK {
			t.Errorf("subdomain %q: API accepts=%v, database accepts=%v — the CHECK in "+
				"0031 and domain.SubdomainPattern have drifted", candidate, apiOK, dbOK)
		}
	}
}
